Adopt Sendable in the Task APIs

DougGregor opened the pull request:
Rework `Sendable` checking to be completely based on "missing"
conformances, so that we can individually diagnose missing `Sendable`
conformances based on both the module in which the conformance check
happened as well as where the type was declared. The basic rules here
are to only diagnose if the non-`Sendable` type either comes from the
current model or from a module that was compiled in a mode that
consistently diagnoses `Sendable`. The latter is true for modules
compiled in Swift 6 mode or that provide `-warn-concurrency` on
the command line. This is tracked by rdar://78269348.

The diagnostics will be an error in Swift 6 or a warning in Swift 5.x.

Use this to adopt `Sendable` in the Task APIs without breaking code, 
tracked by rdar://77441933.

DougGregor committed:
Implement customizable Sendable conformance diagnostics.

Rework Sendable checking to be completely based on "missing"
conformances, so that we can individually diagnose missing Sendable
conformances based on both the module in which the conformance check
happened as well as where the type was declared. The basic rules here
are to only diagnose if either the module where the non-Sendable type
was declared or the module where it was checked was compiled with a
mode that consistently diagnoses `Sendable`, either by virtue of
being Swift 6 or because `-warn-concurrency` was provided on the
command line. And have that diagnostic be an error in Swift 6 or
warning in Swift 5.x.

There is much tuning to be done here.

DougGregor committed:
Delay Codable synthesis notes until after the original diagnostic.

The notes that go along with Codable synthesis were being emitted before
the actual "type does not conform" diagnostic, so they would be associated
with a prior diagnostic... or dropped on the floor. Queue up the
synthesis notes and emit them after the "type does not conform"
diagnostic.

DougGregor committed:
Adopt Sendable in the Task APIs

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
Welp, looks like I need to fix that compile error.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

karwa commented on test/Concurrency/concurrent_value_inference.swift:
```diff
@@ -10,7 +10,7 @@ struct S1 {
 }
 
-enum E1 {
+enum E1 { // expected-note{{add conformance of enum 'E1' to the 'Sendable' protocol}}
```
This message is a bit strange. We don’t add conformance _of_ types _to_ protocols. Might I suggest:

> add conformance to the ‘Sendable’ protocol to the enum ‘E1’.

or even

> add ‘Sendable’ conformance to the enum ‘E1’.

DougGregor replied:
We use the "consider" wording elsewhere, so I'm going to try out

```
consider making enum 'E1' conform to the 'Sendable' protocol 
```

DougGregor committed:
Use a remark, rather than a disembodied note, for constant evaluator

DougGregor committed:
Improve diagnostic wording

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[SE-0302] Crank up `Sendable` to also diagnose in-module missing conformances.

Extend the diagnostics for `Sendable` conformances to always diagnose
missing `Sendable` conformances for nominal types that are within the
same module. The intuition here is that if the type is in the same
module, it can be updated and evaluated at the same time as code
requiring the `Sendable` conformance is introduced.

Another part of rdar://78269348.

DougGregor committed:
[SE-0311] Add Sendable requirement to TaskLocal.

This is part of SE-0311 that was not implemented.
Fixes rdar://77441933.

DougGregor commented:
@swift-ci please smoke test
